File: pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/endpoints/objects_api/uuid/GetUUIDMetadataImpl.java
package com.pubnub.internal.endpoints.objects_api.uuid;

import com.pubnub.api.endpoints.objects_api.uuid.GetUUIDMetadata;
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction;
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction;
import com.pubnub.api.models.consumer.objects_api.uuid.PNGetUUIDMetadataResult;
import com.pubnub.api.models.consumer.objects_api.uuid.PNUUIDMetadata;
import com.pubnub.internal.EndpointInterface;
import com.pubnub.internal.PubNubCore;
import com.pubnub.internal.endpoints.DelegatingEndpoint;
import com.pubnub.internal.models.consumer.objects.uuid.PNUUIDMetadataResult;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.jetbrains.annotations.NotNull;

@Setter
@Accessors(chain = true, fluent = true)
public class GetUUIDMetadataImpl extends DelegatingEndpoint<PNUUIDMetadataResult, PNGetUUIDMetadataResult> implements GetUUIDMetadata {

    private String uuid;
    private boolean includeCustom;

    public GetUUIDMetadataImpl(final PubNubCore pubnub) {
        super(pubnub);
    }

    @Override
    @NotNull
    protected EndpointInterface<PNUUIDMetadataResult> createAction() {
        return pubnub.getUUIDMetadata(
                uuid,
                includeCustom
        );
    }

    @NotNull
    @Override
    protected ExtendedRemoteAction<PNGetUUIDMetadataResult> mapResult(@NotNull ExtendedRemoteAction<PNUUIDMetadataResult> action) {
        return new MappingRemoteAction<>(action, result -> new PNGetUUIDMetadataResult(
                result.getStatus(),
                PNUUIDMetadata.from(result.getData())
        ));
    }
}

File: pubnub-kotlin/src/main/kotlin/com/pubnub/internal/kotlin/endpoints/objects/uuid/GetUUIDMetadataImpl.kt
package com.pubnub.internal.kotlin.endpoints.objects.uuid

import com.pubnub.api.DelegatingEndpoint
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction
import com.pubnub.api.models.consumer.objects.uuid.PNUUIDMetadataResult
import com.pubnub.internal.PubNubImpl
import com.pubnub.internal.endpoints.objects.uuid.GetUUIDMetadata
import com.pubnub.internal.endpoints.objects.uuid.IGetUUIDMetadata
import com.pubnub.internal.models.consumer.objects.uuid.PNUUIDMetadataResult as PNUUIDMetadataResultInternal

/**
 * @see [PubNubImpl.getUUIDMetadata]
 */
class GetUUIDMetadataImpl internal constructor(getUUIDMetadata: GetUUIDMetadata) :
    DelegatingEndpoint<PNUUIDMetadataResult, PNUUIDMetadataResultInternal>(getUUIDMetadata),
    IGetUUIDMetadata by getUUIDMetadata,
    com.pubnub.api.endpoints.objects.uuid.GetUUIDMetadata {
    override fun convertAction(remoteAction: ExtendedRemoteAction<PNUUIDMetadataResultInternal>): ExtendedRemoteAction<PNUUIDMetadataResult> {
        return MappingRemoteAction(remoteAction, PNUUIDMetadataResult.Companion::from)
    }
}
